bam-index/src/main/java/org/victorchang/QnameSearcher.java

```diff
@@ -29,83 +33,103 @@ public QnameSearcher(KeyPointerReader keyPointerReader, BamRecordReader recordRe
         this.handler = handler;
     }
 
-    public int search(Path bamFile, Path indexFolder, String qname) throws IOException {
-        Path pathLevel1 = indexFolder.resolve("qname.1");
-        FileChannel channelLevel1 = FileChannel.open(pathLevel1, READ);
-        InputStream inputStreamLevel1 = Channels.newInputStream(channelLevel1);
-
-        byte[] input = Ascii7Coder.INSTANCE.encode(qname);
-        Iterable<KeyPointer> indexLevel1 = () -> keyPointerReader.read(inputStreamLevel1).iterator();
-
-        KeyPointer key = new KeyPointer(0, input, input.length);
-        KeyPointer start = key;
-        for (KeyPointer x : indexLevel1) {
-            if (x.compareTo(key) >= 0) {
-                break;
-            }
-            start = x;
+    public int search(Path bamFile, Path indexFolder, Set<String> qnames) throws IOException {
+        final List<Long> pointersForQname = getPointersForQname(indexFolder, qnames);
+        for (Long pointer : pointersForQname) {
+            recordReader.read(bamFile, pointer, handler);
         }
-        inputStreamLevel1.close();
-
-        Path pathLevel0 = indexFolder.resolve("qname.0");
-        FileChannel channelLevel0 = FileChannel.open(pathLevel0, READ);
-        long coffset = PointerPacker.INSTANCE.unpackCompressedOffset(start.getPointer());
-        int uoffset = PointerPacker.INSTANCE.unpackUnCompressedOffset(start.getPointer());
+        return pointersForQname.size();
+    }
 
-        if (coffset >= channelLevel0.size()) {
-            return 0;
-        }
-        channelLevel0.position(coffset);
-        InputStream inputStreamLevel0 = Channels.newInputStream(channelLevel0);
-
-        int found = 0;
-        Iterable<KeyPointer> indexLevel0 =  () -> keyPointerReader.read(inputStreamLevel0, uoffset).iterator();
-        for (KeyPointer x : indexLevel0) {
-            if (Arrays.equals(x.getKey(), input)) {
-                recordReader.read(bamFile, x.getPointer(), handler);
-                found++;
-            }
-            if (Arrays.compareUnsigned(x.getKey(), input) > 0) {
-                break;
-            }
-        }
+    List<Long> getPointersForQname(Path indexFolder, Set<String> qnames) throws IOException {
+        final Map<byte[], KeyPointer> qnameToPointer = getQnameToPointerMap(indexFolder, qnames);
 
-        inputStreamLevel0.close();
+        // Map from a pointer -> set of qnames in that level
+        final Map<Long, Set<byte[]>> pointerToQname = qnameToPointer.entrySet().stream()
```

Comment:
I don't think we need it Map from pointer to a set of qname because each pointer points to a record that has only a single qname.

Reply:
The pointer here is the start pointer in the level-1 index. They can have multiple qnames in the same block.